### ISSUE TITLE
Possible data loss in the JACK backend.

### ISSUE THREAD

**Issue by keinstein** [thestk/rtmidi#24]
The output ringbuffer is not flushed when the jack client is closed. If the JACK connection thread did not deliver all data, the remaining data is lost.

**Comment by garyscavone** [comment]
Can someone propose a pull request?

**Comment by radarsat1** [comment]
I'm working on it.  I think the destructor has to wait (with a timeout) on jackProcessOut() to completely read the ringbuffer, but I haven't verified a solution yet.

**Comment by radarsat1** [comment]
I tested this pretty thoroughly on my local machine https://github.com/thestk/rtmidi/commit/ee4643b76397b451761137290f2873bc9cc636f1

to make sure all messages got through. Without this patch the last message from midiout is dropped.  I used a pair of POSIX semaphores to solve the problem, one with which jackProcessOut uses sem_trywait to indicate that it needs to sem_post to the other semaphore, and closePort sem_posts to the first semaphore and then waits on the second with a 1-second timeout.  I hope this is a good approach.. I am not sure it is conventional, but it seems to work and is non-blocking.  Any comments welcome.
